### PR TITLE
fix(android): Fix auth flow and callback thread safety, and pass `fd` through FFI

### PR DESCRIPTION
**Pull request by jamilbk**
* Refactor sharedPreferences to only save the AccountId
* Update TeamId -> AccountId to match naming elsewhere
* Update JWT -> Token to avoid confusion; this token is **not** a valid JWT and should be treated as an opaque token
* Update FFI `connect` to accept an optional file descriptor (int32) as a first argument. This seemed to be the most straightforward way to pass it to the tunnel stack. Retrieving it via callback is another option, but retrieving return vars with the `jni` was more complex. We could have used a similar approach that we did in the Apple client (enumerating all fd's in the `new()` function until we found ours) but this approach is [explicitly documented/recommended](https://developer.android.com/reference/android/net/VpnService.Builder#establish()) by the Android docs so I figured it's not likely to break.

Additionally, there was a thread safety bug in the recent JNI callback implementation that consistently crashed the VM with `JNI DETECTED ERROR IN APPLICATION: use of invalid jobject...`. The fix was to use `GlobalRef` which has the explicit purpose of outliving the `JNIEnv` lifetime so that no `static` lifetimes need to be used.

### CODE REVIEW

File: kotlin/android/app/build.gradle
Comment (jamilbk) [inline kotlin/android/app/build.gradle]: This is the host's IP as seen by the android debugging environment.

File: kotlin/android/app/build.gradle
Comment (jamilbk) [inline kotlin/android/app/build.gradle]: @pratikvelani You should now have an account on staging you can use to sign in with the magic link method, which has real Gateways, and a live connlib working with it. Use `pratik@spark.dev` to sign in.

With this change you no longer need to run `PortalMock`.

File: kotlin/android/app/src/main/AndroidManifest.xml
Comment (jamilbk) [inline kotlin/android/app/src/main/AndroidManifest.xml:14]:
```suggestion
            <action android:name="android.support.customtabs.action.CustomTabsService" />
```

This fixed a crash I was hitting with CustomTabs.

File: kotlin/android/app/src/main/AndroidManifest.xml
Comment (jamilbk) [inline kotlin/android/app/src/main/AndroidManifest.xml:61]:
```suggestion
                <data android:scheme="https" android:host="${hostName}" android:path="/handle_client_auth_callback" />
```

Without the `path` specifier here, the AppLink would intercept the initial `sign_in` path, preventing the user from actually signing in.

File: kotlin/android/app/src/main/java/dev/firezone/android/features/applink/ui/AppLinkHandlerActivity.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/features/applink/ui/AppLinkHandlerActivity.kt:38]: @pratikvelani This is where I left off. We need to show the session fragment instead of the `AUTH Complete` TextView here.

File: kotlin/android/app/src/main/java/dev/firezone/android/features/session/ui/SessionViewModel.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/features/session/ui/SessionViewModel.kt]:
```suggestion
```

File: kotlin/android/app/src/main/java/dev/firezone/android/features/session/backend/SessionManager.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/features/session/backend/SessionManager.kt]: @conectado Here is where it would be convenient to pass in the file descriptor in a single line

File: kotlin/android/app/src/main/java/dev/firezone/android/util/CustomTabsHelper.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/util/CustomTabsHelper.kt:1]: @pratikvelani This was taken from the CustomTabs demo repo. I think we should support other browsers that the user may have set as default (e.g. Firefox)

**Comment by jamilbk** [comment]
@pratikvelani I'm not sure what issue you're hitting with the connlib callbacks, because the Android application had a number of authentication flow issues preventing me from getting to `Session.connect` call to replicate what you are seeing.

This should fix most of them. Please let me know if you're able to get an _actual_ auth flow working against staging, and we can debug the `connlib` callbacks then.

I'd like to avoid using Mocks since we have a working staging environment now.

The local Docker dev environment won't be easy to test against due to all the NAT'ing involved from Android emulator to docker network...

File: elixir/apps/web/lib/web/endpoint.ex
Comment (jamilbk) [inline elixir/apps/web/lib/web/endpoint.ex:42]: @andrewdryga this fixed liveview issues in development when accessing the portal by my Mac's local LAN IP. Unsure if we want this on `prod` however?

File: elixir/config/runtime.exs
Comment (jamilbk) [inline elixir/config/runtime.exs:87]: @AndrewDryga These actually don't change between environments. The App Link is only intercepted for `app.firez.one`

File: kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt
Comment (pratikvelani) [inline kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt:18]: @jamilbk I was implementing a separate SettingsFragment, as it would have some UI for the user to delete the token and in debug mode a UI for us to view app logs.

**Review by pratikvelani — APPROVED** [review]

File: kotlin/android/app/build.gradle
Comment (jamilbk) [reply kotlin/android/app/build.gradle]: Nevermind, using `app.firez.one` since testing locally with Docker is problematic

File: kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt
Comment (jamilbk) [reply kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt:18]: Ah, but wouldn't `Sign out` accomplish deleting the token? Wiping the app's data would too right? Confused why that would be a dedicated action in the Settings fragment.

A debug log viewer in the SettingsFragment sounds good. Do you want to merge this and add your changes into it, or the other way around?

The main thing here was renaming `OnboardingFragment` since that's not really what this view is.

File: elixir/apps/web/lib/web/endpoint.ex
Comment (AndrewDryga) [reply elixir/apps/web/lib/web/endpoint.ex:42]: That's actually a good catch and we do want it on prod, thank you

File: kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt
Comment (pratikvelani) [reply kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt:18]: I want to move the `Sign out` button to the settings screen too. We can merge this PR, I will push a separate PR for it.

**Comment by jamilbk** [comment]
Adding @thomaseizinger as another pair of eyes on the Rust JNI stuff. It seems to be working, but I don't trust myself:

```
2023-08-21 20:43:06.771 12432-12489 connlib                 dev.firezone.android                 D  jni::wrapper::java_vm::vm: Attached thread tokio-runtime-worker (ThreadId(1)). 1 threads attached
2023-08-21 20:43:06.771 12432-12489 TunnelCallbacks         dev.firezone.android                 D  onError: Portal connection error: IO error: invalid peer certificate: UnknownIssuer
2023-08-21 20:43:06.771 12432-12489 connlib                 dev.firezone.android                 D  jni::wrapper::java_vm::vm: Detached thread tokio-runtime-worker (ThreadId(1)). 0 threads remain attached
2023-08-21 20:43:47.966 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::hs: No cached session for DnsName("app.firez.one")
2023-08-21 20:43:47.967 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::hs: Not resuming any session
2023-08-21 20:43:47.978 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::hs: Using ciphersuite TLS13_AES_256_GCM_SHA384
2023-08-21 20:43:47.978 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::tls13: Not resuming
2023-08-21 20:43:47.979 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::tls13: TLS1.3 encrypted extensions: []
2023-08-21 20:43:47.979 12432-12490 connlib                 dev.firezone.android                 D  rustls::client::hs: ALPN protocol is None
2023-08-21 20:43:47.979 12432-12490 connlib                 dev.firezone.android                 D  jni::wrapper::java_vm::vm: Attached thread tokio-runtime-worker (ThreadId(2)). 1 threads attached
2023-08-21 20:43:47.980 12432-12490 TunnelCallbacks         dev.firezone.android                 D  onError: Portal connection error: IO error: invalid peer certificate: UnknownIssuer
2023-08-21 20:43:47.980 12432-12490 connlib                 dev.firezone.android                 D  jni::wrapper::java_vm::vm: Detached thread tokio-runtime-worker (ThreadId(2)). 0 threads remain attached
```

Now to fix the cert error...

File: kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt
Comment (jamilbk) [reply kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt:18]: Sounds good. This is one is already large enough.

**Review by pratikvelani — APPROVED** [review]

**Comment by jamilbk** [comment]
@conectado I think this is ready to go once https://github.com/firezone/webrtc/pull/1 is approved and merged. I'll open another PR with the refactoring we discussed earlier today to get the `fd` callback and tunnel interface addresses corrected.

File: rust/connlib/clients/android/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/android/src/lib.rs:328]: Just a note to myself in the future:

The problem here was that an owned `JObject` can never have an `'static` lifetime since by default `JObject`s have a lifetime of the frame that's calling the ffi function and the only way to obtain an static one is through a `GlobalRef` which only implements `AsRef<JObject<'static>>`.

File: rust/connlib/clients/android/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/android/src/lib.rs]:
```suggestion
let Ok(callback_handler) = env.new_global_ref(callback_handler) else { return std::ptr::null() };
```

File: rust/connlib/clients/android/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/android/src/lib.rs]: this is still `unsafe` since the pointers here might be unvalid.

Comment (conectado) [reply rust/connlib/clients/android/src/lib.rs]: Also now we need to assume that `fd` is a vaild file descriptor.

File: rust/connlib/clients/android/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/android/src/lib.rs]: This can still be a `&JObject`, it doesn't implement neither `Send` nor `Sync` so we can be sure that it's not moved to another thread.

File: rust/connlib/clients/android/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/android/src/lib.rs]: We do nothing with this error so we can remove it.

File: rust/connlib/libs/tunnel/src/lib.rs
Comment (conectado) [inline rust/connlib/libs/tunnel/src/lib.rs:511]: But it could still happen that it changes, if that happens this could panic.

File: rust/connlib/libs/gateway/src/control.rs
Comment (conectado) [inline rust/connlib/libs/gateway/src/control.rs:147]: We are going to change it later so it doesn't matter but I think it's better if we expect here a `RawFd`.

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (conectado) [inline rust/connlib/libs/tunnel/src/tun_android.rs:56]: we still should find the actual mtu or at least have the callbacks tell us the mtu that it used at some point, since if it configured a great MTU we could read packets partially.

Comment (conectado) [reply rust/connlib/libs/tunnel/src/tun_android.rs:56]: also note that this should be the wireguard's interface mtu not the physical interface, each peer could use a different physical interface and it can also change.

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (conectado) [inline rust/connlib/libs/tunnel/src/tun_android.rs:39]: This is actually unsafe since `fd` might not be valid and we assume it's vaild in the `read`/`write` implementations. Since we are going to change how we obtain the `fd` it's not that important right now but we need to remember that for the refactor.

File: rust/connlib/libs/common/Cargo.toml
Comment (conectado) [inline rust/connlib/libs/common/Cargo.toml:35]: I think we should enable the [`log` feature](https://docs.rs/tracing/latest/tracing/#crate-feature-flags) from the tracing crate instead. then we can just drop `log` entirely and `tracing` will automatically emit `log`'s logs if we are on android.

**Review by conectado — DISMISSED** [review]
Looks good! left some comments that'd be good if we can fix before merging.

File: rust/connlib/libs/tunnel/src/lib.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/lib.rs:511]: I see. Pardon my noob question, but won't adding syscalls to the hotpath take up a lot of CPU time / overhead?

I ask because WireGuard-go [stores the MTU on the device struct](https://github.com/WireGuard/wireguard-go/blob/469159ecf7d108dc1d44f8789dedbb463b7bd3a6/device/device.go#L86), and boringtun [does something similar](https://github.com/cloudflare/boringtun/blob/e1d6360d6ab4529fc942a078e4c54df107abe2ba/boringtun/src/device/mod.rs#L382).

If so, I think it's an acceptable edge case for this to possibly crash if an admin changes the tunnel interface MTU while it's up.

Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/lib.rs:511]: From there, if it is crashing and we do find this to be an unacceptable UX in the wild, we could refresh it periodically, or maybe it's possible to subscribe to the MTU set syscall?

File: rust/connlib/libs/gateway/src/control.rs
Comment (jamilbk) [reply rust/connlib/libs/gateway/src/control.rs:147]: Yeah, this thought was already crossing my mind as I wrote this, so I'll note it down for when we change it!

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/tun_android.rs:56]: It's a little involved in Android since we first have to retrieve the interface name with `TUNGETIFF`. But I'll take that on as an exercise for the reader.

File: rust/connlib/libs/common/Cargo.toml
Comment (jamilbk) [reply rust/connlib/libs/common/Cargo.toml:35]: Nice, it looks like we can also consume the log events from our dependencies and consume those in tracing as well with https://docs.rs/tracing/latest/tracing/#consuming-log-records

Perhaps I'll save that change for another PR?

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/tun_android.rs:39]: Added

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/tun_android.rs:56]: I'll save this for the next PR if that's alright

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Removed

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Fixed

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Fixed

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Fixed

**Comment by jamilbk** [comment]
@conectado Addressed the main things, saving the bigger items for another PR if that's alright

File: rust/connlib/libs/tunnel/src/device_channel_unix.rs
Comment (pratikvelani) [inline rust/connlib/libs/tunnel/src/device_channel_unix.rs:64]:
```suggestion
    let dev = Arc::new(unsafe { IfaceDevice::new(fd).await?.set_non_blocking()? });
```

Fixes CI build issue
```rust
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
  --> connlib/libs/tunnel/src/device_channel_unix.rs:64:24
   |
64 |     let dev = Arc::new(IfaceDevice::new(fd).await?.set_non_blocking()?);
   |                        ^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior
```

**Review by AndrewDryga — APPROVED** [review]

File: rust/connlib/libs/tunnel/src/device_channel_unix.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/device_channel_unix.rs:64]: Fixed by removing`unsafe` from the function definition.

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/tun_android.rs:39]: Aaaand removed so this will compile

File: rust/connlib/libs/tunnel/src/lib.rs
Comment (conectado) [reply rust/connlib/libs/tunnel/src/lib.rs:511]: Now that I thought about this, I think that the `read` call will not fail if the buffer is small it will just read a mangled packet(the rest of this loop expects a full packet).

So we can just do as boringtun does that it reads the mtu once and updates it periodically. Still calling a method here is fine, the `DeviceChannel` can store the mtu.

File: rust/phoenix-channel/Cargo.toml
Comment (conectado) [inline rust/phoenix-channel/Cargo.toml]:
```suggestion
```

`phoenix-channel` is only used by the relay, it doesn't need `log`.

**Review by conectado — APPROVED** [review]
Left a few extra non-blocking comments.

File: rust/connlib/libs/tunnel/src/lib.rs
Comment (jamilbk) [reply rust/connlib/libs/tunnel/src/lib.rs:511]: Gotcha, that makes sense. Thanks for clarifying! Glad it wasn't a stupid question.

**Comment by jamilbk** [comment]
Nice, thanks GitHub. There are unresolved convos we can resolve because the commits don't exist anymore (I force-pushed). Going to close and reopen to try and resolve.

**Pull request merged**
